File: aether-bridge/messages.py
"""Wire-format message builders shared by bridge, simulated scanner, and aggregator.

These functions define the JSON contract sent to WebSocket clients (dashboard).
Field names here are a locked contract - do not rename fields without updating
every consumer.
"""

from datetime import datetime


def _now_hms() -> str:
    return datetime.now().strftime("%H:%M:%S")


def build_reading_message(
    scanner: str, name: str, raw_rssi: float, smoothed_rssi: float, last_seen_ms: int
) -> dict:
    return {
        "type": "reading",
        "scanner": scanner,
        "name": name,
        "rssi": raw_rssi,
        "smoothedRssi": smoothed_rssi,
        "lastSeenMs": last_seen_ms,
        "ts": _now_hms(),
    }


def build_lost_message(scanner: str, name: str) -> dict:
    return {
        "type": "lost",
        "scanner": scanner,
        "name": name,
        "ts": _now_hms(),
    }


def build_election_message(
    owner: str | None,
    tick: int,
    scanners: list[dict],
    last_handoff: dict | None,
    wake_outcome: dict | None,
) -> dict:
    """Build the election broadcast message.

    Locked schema (dashboard consumes this directly):
        {"type": "election", "owner": "PC", "tick": 4831, "ts": "14:32:41",
         "scanners": [{"id": "PC", "rssi": -58.2, "smoothedRssi": -59.1,
                        "lastSeenMs": 340, "present": true}, ...],
         "lastHandoff": {"from": "SIM-A", "to": "PC", "atTick": 4821,
                          "ts": "14:32:15"},
         "wakeOutcome": {"requestedAtTick": 4830, "ts": "14:32:41",
                          "owner": "PC",
                          "results": [{"id": "PC", "outcome": "ACCEPTED"},
                                      {"id": "SIM-A", "outcome": "SUPPRESSED"}]}}

    `scanners` must be supplied in stable peers-list order, one entry per
    configured peer regardless of liveness. `last_handoff` and `wake_outcome`
    are passed through as-is (callers control the one-shot / most-recent-only
    semantics); this function performs no logic beyond envelope assembly.
    """
    return {
        "type": "election",
        "owner": owner,
        "tick": tick,
        "ts": _now_hms(),
        "scanners": scanners,
        "lastHandoff": last_handoff,
        "wakeOutcome": wake_outcome,
    }


def build_conversation_message(
    transcript: list[dict],
    utterance: dict | None,
    speaking_scanner: str | None,
    phase: str,
    phase_from: str | None,
    phase_to: str | None,
    conversation_event: dict | None,
) -> dict:
    """Build the conversation broadcast message.

    Sent as a separate JSON message on the same WebSocket as the election
    broadcast (immediately after it). Locked schema (dashboard consumes this
    directly):

        {"type": "conversation",
         "transcript": [{"id": 1, "scanner": "SIM-A", "role": "assistant",
                         "text": "...", "ts": "14:32:41"}, ...],
         "utterance": {"text": "...", "audioBase64": "data:audio/mp3;base64,...",
                       "durationMs": 2400, "offsetMs": 0,
                       "isSynthetic": false} | null,
         "speakingScanner": "SIM-A" | null,
         "phase": "IDLE" | "PREPARE" | "TRANSFER" | "CONFIRM" | "RELEASE",
         "phaseFrom": "SIM-A" | null,
         "phaseTo": "SIM-B" | null,
         "conversationEvent": {"phase": "PREPARE", "fromScanner": "SIM-A",
                               "toScanner": "SIM-B", "atTick": 4830} | null}

    `conversationEvent` is one-shot (caller clears it after one broadcast,
    mirroring wakeOutcome's semantics) and marks a phase transition for the
    dashboard to react to (e.g. pause audio on TRANSFER, resume on RELEASE).
    """
    return {
        "type": "conversation",
        "transcript": transcript,
        "utterance": utterance,
        "speakingScanner": speaking_scanner,
        "phase": phase,
        "phaseFrom": phase_from,
        "phaseTo": phase_to,
        "conversationEvent": conversation_event,
    }

File: aether-android/shared/src/main/kotlin/com/aether/shared/Messages.kt
package com.aether.shared

import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.time.LocalTime
import java.time.format.DateTimeFormatter

/**
 * Wire-format message builders mirroring `aether-bridge/messages.py`.
 *
 * Field names here are a locked contract - do not rename fields without
 * updating every consumer (`aggregator.py`, dashboard). This module only
 * implements the message types an Android scanner produces: `reading` and
 * `lost` (§ PRD.md acceptance criteria - no changes to messages.py/aggregator.py
 * required, since Android is just another producer of the same wire format).
 */
public object Messages {

    private val HMS_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss")

    private fun nowHms(): String = LocalTime.now().format(HMS_FORMATTER)

    /**
     * Builds the JSON string for a `reading` message, matching
     * `messages.py`'s `build_reading_message` field-for-field:
     * `{"type": "reading", "scanner", "name", "rssi", "smoothedRssi", "lastSeenMs", "ts"}`.
     */
    public fun buildReadingMessage(
        scanner: String,
        name: String,
        rawRssi: Double,
        smoothedRssi: Double,
        lastSeenMs: Long,
    ): String {
        return buildJsonObject {
            put("type", JsonPrimitive("reading"))
            put("scanner", JsonPrimitive(scanner))
            put("name", JsonPrimitive(name))
            put("rssi", JsonPrimitive(rawRssi))
            put("smoothedRssi", JsonPrimitive(smoothedRssi))
            put("lastSeenMs", JsonPrimitive(lastSeenMs))
            put("ts", JsonPrimitive(nowHms()))
        }.toString()
    }

    /**
     * Builds the JSON string for a `lost` message, matching
     * `messages.py`'s `build_lost_message` field-for-field:
     * `{"type": "lost", "scanner", "name", "ts"}`.
     */
    public fun buildLostMessage(scanner: String, name: String): String {
        return buildJsonObject {
            put("type", JsonPrimitive("lost"))
            put("scanner", JsonPrimitive(scanner))
            put("name", JsonPrimitive(name))
            put("ts", JsonPrimitive(nowHms()))
        }.toString()
    }
}
